dali/c_api/c_api.cc

```diff
@@ -70,6 +73,59 @@ void daliPrefetchSeparate(daliPipelineHandle* pipe_handle,
   }
 }
 
+void daliSetExternalInput(daliPipelineHandle* pipe_handle, const char* name, device_type_t device,
+                          const void* data_ptr, dali_data_type_t data_type, const int64_t* shape,
+                          int sample_dim, const char* layout_str) {
+  DALI_ENFORCE(device == CPU, "GPU data cannot be passed as external source");
+  dali::Pipeline* pipeline = reinterpret_cast<dali::Pipeline*>(pipe_handle->pipe);
+  std::vector<int64_t> shapes_tmp(shape, shape + sample_dim * pipeline->batch_size());
+  dali::TensorListShape<> shapes(std::move(shapes_tmp), pipeline->batch_size(), sample_dim);
+  std::string source_name(name);
+  dali::TensorLayout layout{};
+  if (layout_str != nullptr) {
+    layout = dali::TensorLayout(layout_str);
+  }
+  dali::TensorList<dali::CPUBackend> data;
+  const auto& type_info = dali::TypeTable::GetTypeInfo(static_cast<dali::DALIDataType>(data_type));
+  auto elem_sizeof = type_info.size();
+  // We cast away the const from data_ptr, as there is no other way of passing it to the
+  // TensorList, as we must also set the shape and type metadata.
+  // It is passed further as const TensorList, so it's data cannot be modified.
+  data.ShareData(const_cast<void*>(data_ptr), shapes.num_elements() * elem_sizeof);
+  data.Resize(shapes, type_info);
+  data.SetLayout(layout);
+  const auto& cdata = data;
+  pipeline->SetExternalInput(source_name, cdata);
+}
+
+void daliSetExternalInputTensors(daliPipelineHandle* pipe_handle, const char* name,
+                                 device_type_t device, const void* const* data_ptr,
+                                 dali_data_type_t data_type, const int64_t* shape,
+                                 int64_t sample_dim, const char* layout_str) {
+  DALI_ENFORCE(device == CPU, "GPU data cannot be passed as external source");
+  dali::Pipeline* pipeline = reinterpret_cast<dali::Pipeline*>(pipe_handle->pipe);
+  std::vector<int64_t> shapes_tmp(shape, shape + sample_dim * pipeline->batch_size());
+  dali::TensorListShape<> shapes(std::move(shapes_tmp), pipeline->batch_size(), sample_dim);
+  std::string source_name(name);
+  dali::TensorLayout layout{};
+  if (layout_str != nullptr) {
+    layout = dali::TensorLayout(layout_str);
+  }
+  std::vector<dali::Tensor<dali::CPUBackend>> data(pipeline->batch_size());
+  const auto& type_info = dali::TypeTable::GetTypeInfo(static_cast<dali::DALIDataType>(data_type));
+  auto elem_sizeof = type_info.size();
+  for (int i = 0; i < pipeline->batch_size(); i++) {
+    // We cast away the const from data_ptr, as there is no other way of passing it to the
+    // Tensor as we must also set the shape and type metadata.
+    // The vector that we pass to pipeline is const.
+    data[i].ShareData(const_cast<void*>(data_ptr[i]), shapes[i].num_elements() * elem_sizeof);
+    data[i].Resize(shapes[i], type_info);
+    data[i].SetLayout(layout);
+  }
+  const auto& cdata = data;
+  pipeline->SetExternalInput(source_name, cdata);
```

Comment:
likewise - why have the extra `source_name` variable?

Reply:
Does it make any difference? It will still create a temporary.